tools/schemapi/schemapi.py

```diff
@@ -1393,45 +1699,50 @@ def from_dict(
         """Construct an object from a dict representation."""
         target_tp: Any
         current_schema: dict[str, Any]
+        hash_schema: int
         if isinstance(dct, SchemaBase):
             return dct
         elif tp is not None:
             current_schema = tp._schema
+            hash_schema = _hash_schema(current_schema)
             root_schema: dict[str, Any] = rootschema or tp._rootschema or current_schema
             target_tp = tp
         elif schema is not None:
-            # If there are multiple matches, we use the first one in the dict.
-            # Our class dict is constructed breadth-first from top to bottom,
-            # so the first class that matches is the most general match.
             current_schema = schema
+            hash_schema = _hash_schema(current_schema)
             root_schema = rootschema or current_schema
-            matches = self.class_dict[self.hash_schema(current_schema)]
-            target_tp = matches[0] if matches else default_class
+            matches = cls.hash_tps[hash_schema]
+            target_tp = next(iter(matches), default_class)
```

Comment:
FIXME: This is the slow branch.

- Improving the performance of the `tp` one has no effect
  - I tried storing the hash value as a `ClassVar` in `SchemaBase`
  - Conceptually made sense, but this branch is triggered much more often
- Every recursive `from_dict` call that isn't solved hits this